packages/frontend-shared/src/utils/getUrlWithParams.ts

```diff
@@ -7,14 +7,29 @@ export type LinkWithParams = {
 
 export const getUrlWithParams = (link: LinkWithParams) => {
   let result = link.url
-  const hasUtmParams = Object.keys(link.params).some((param) => param.startsWith('utm_'))
 
-  if (hasUtmParams) {
-    link.params.utm_source = getUtmSource()
-  }
+  if (Object.keys(link.params).length > 0) {
+    const hasUtmParams = Object.keys(link.params).some((param) => param.startsWith('utm_'))
+
+    if (hasUtmParams) {
+      link.params.utm_source = getUtmSource()
+    }
+
+    let url: string
+    let searchParams: URLSearchParams
+
+    if (link.url.includes('?')) {
+      // If input URL already includes params we should preserve them
+      url = link.url.substring(0, link.url.indexOf('?'))
+      searchParams = new URLSearchParams(link.url.substring(link.url.indexOf('?')))
+    } else {
+      url = link.url
+      searchParams = new URLSearchParams()
+    }
+
+    Object.entries(link.params).forEach(([key, value]) => searchParams.append(key, value))
```

Comment:
Looks like this can be simplified by using `new URL(...)`, which will preserve the searchParams if already provided.

```suggestion
    const url = new URL(link.url)

    Object.entries(link.params).forEach(([key, value]) => url.searchParams.append(key, value))

    result = url.href
```

Reply:
I started out with that approach, but hit a hurdle since it doesn't handle "partial" URLs like `./my/path` unless we manually provide a base URL as a separate argument. I wasn't sure whether it was safe to assume a "localhost"-relative path in all circumstances so I took the manual parsing approach to leave the relative-pathing decisions up to the browser when the link was resolved

Reply:
Ahh makes sense, your approach seems good then.